### PR TITLE
Independent RMH implementation

### PR DESCRIPTION
**Pull request by ciguaran**
This is related to https://github.com/blackjax-devs/blackjax/issues/245

.Adds Independent RMH, adds a simple test in test_sampling.py
.Refactors RMH so that it takes a proposal distribution, vs the current behavior that takes a proposal for the movement wrt the current state. Maintains backwards compatibility. 

- [x] We should be able to understand what the PR does from its title only;  
- [x] There is a high-level description of the changes;
- [x] There are links to *all* the relevant issues, discussions and PRs;
- [x] The branch is rebased on the latest `main` commit;
- [x] Commit messages follow these [guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html);
- [x] `pre-commit` is installed and configured on your machine, and you ran it before opening the PR;
- [x] There are tests covering the changes;
- [x] The doc is up-to-date;
- [ ] [NO, check issue blackjax-devs/blackjax/issues/245 for an explanation] If I add a sampler I added/updated related [examples](https://github.com/blackjax-devs/blackjax/tree/main/examples)

@rlouf @aloctavodia

### CODE REVIEW

File: tests/test_sampling.py [hunk 1]
```diff
@@ -208,6 +208,20 @@ def test_linear_regression(self):
         _ = sgld.step(rng_key, init_state, data_batch)
 
 
+irmh_test_case = {
```
Comment (ciguaran) [inline tests/test_sampling.py:211]: I extracted this in order to have it close to the comment explaining the rationale for the proposal distribution. LMK if other format is preferred.

File: blackjax/mcmc/irmh.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+"""Public API for Rosenbluth-Metropolis-Hastings kernels."""
```
Comment (rlouf) [inline blackjax/mcmc/irmh.py:1]:
```suggestion
"""Public API for the Independent Rosenbluth-Metropolis-Hastings kernels."""
```

File: blackjax/mcmc/irmh.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    """Build an Independent Random Walk Rosenbluth-Metropolis-Hastings kernel. This implies
+    that the proposal distribution does not depend on the particle being mutated.
+
+    Returns
+    -------
+    A kernel that takes a rng_key and a Pytree that contains the current state
+    of the chain and that returns a new state of the chain along with
+    information about the transition.
+    """
+
```
[the remark this answers is not in the document]
Comment (rlouf) [reply blackjax/mcmc/irmh.py:20]: We try to keep this convention throughout the library

File: blackjax/mcmc/irmh.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+from blackjax.mcmc.rmh import RMHInfo, RMHState, rmh_with_proposal_generator
+from blackjax.types import PRNGKey, PyTree
+
+__all__ = ["kernel"]
+
+
+def kernel(proposal_distribution: Callable):
+    """Build an Independent Random Walk Rosenbluth-Metropolis-Hastings kernel. This implies
+    that the proposal distribution does not depend on the particle being mutated.
+
```
Comment (rlouf) [inline blackjax/mcmc/irmh.py:13]: Is there a paper from the literature that you could reference here?

File: blackjax/mcmc/rmh.py [hunk 5]
```diff
@@ -137,7 +154,7 @@ def kernel(rng_key: PRNGKey, state: RMHState) -> Tuple[RMHState, RMHInfo]:
         """Move the chain by one step using the Rosenbluth Metropolis Hastings
         algorithm.
 
-        We temporarilly assume that the proposal distribution is symmetric.
+        We temporally assume that the proposal distribution is symmetric.
```
Comment (rlouf) [inline blackjax/mcmc/rmh.py:157]:
```suggestion
        We temporarily assume that the proposal distribution is symmetric.
```

File: blackjax/mcmc/rmh.py [hunk 6]
```diff
@@ -94,7 +94,7 @@ def one_step(
 
 def rmh(
     logprob_fn: Callable,
-    proposal_generator: Callable,
```
Comment (rlouf) [inline blackjax/mcmc/rmh.py:97]: Good catch, naming was indeed ambiguous

**Comment by rlouf** [comment]
Thank you for the contribution! Can you briefly explain what the difference between RHM and independent RMH is exactly so I can keep reviewing?

**Comment by junpenglao** [comment]
> Thank you for the contribution! Can you briefly explain what the difference between RHM and independent RMH is exactly so I can keep reviewing?

From my understanding, RMH transition is `f(x) -> random.normal(loc=x, scale=scale)` and Independent transition is `g(_) -> random.normal(loc=x, scale=scale)`, in the later case `x` is fixed (e.g., estimate from higher temperature in SMC). The same MH accept-reject still applies in both cases.

**Comment by ciguaran** [comment]
Check Algorithm 2 here : https://arxiv.org/pdf/2008.02455.pdf

File: blackjax/kernels.py [hunk 7]
```diff
@@ -648,6 +649,61 @@ def step_fn(rng_key: PRNGKey, state):
...
+
+        def init_fn(position: PyTree):
+            return cls.init(position, logprob_fn)
+
+        def step_fn(rng_key: PRNGKey, state):
+            return step(rng_key, state, logprob_fn)
+
+        return SamplingAlgorithm(init_fn, step_fn)
+
+
```
[the remark this answers is not in the document]
Comment (ciguaran) [reply blackjax/kernels.py:706]: The thing is that at some point we need users to be able to run something like
`blackjax.adaptive_tempered_smc(
    logprior,
    loglikelihood,
    blackjax.irmh,
    irmh_parameters,
    resampling.systematic,
    target_ess=0.8,
    mcmc_iter=10)`

So they need to be able to reference IRMH in order to run it (even as SMC kernel).  Unless there's a better place for SMC kernels that are only meant to be used in that context.

File: blackjax/mcmc/irmh.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+from blackjax.types import PRNGKey, PyTree
+
+__all__ = ["kernel"]
+
+
+def kernel(proposal_distribution: Callable):
+    """Build an Independent Random Walk Rosenbluth-Metropolis-Hastings kernel. This implies
+    that the proposal distribution does not depend on the particle being mutated.
+    Reference: Algorithm 2 from https://arxiv.org/pdf/2008.02455.pdf
+    Returns
```
Comment (rlouf) [inline blackjax/mcmc/irmh.py:14]: nitpick: linebreak

File: blackjax/mcmc/irmh.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
+"""Public API for the Independent Rosenbluth-Metropolis-Hastings kernels."""
+from typing import Callable, Tuple
+
+from blackjax.mcmc.rmh import RMHInfo, RMHState, rmh_with_proposal_generator
+from blackjax.types import PRNGKey, PyTree
+
+__all__ = ["kernel"]
+
+
+def kernel(proposal_distribution: Callable):
```
Comment (rlouf) [inline blackjax/mcmc/irmh.py:10]: Can you add a return type here?

File: blackjax/kernels.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (rlouf) [reply blackjax/kernels.py:706]: I see. We didn't propagate recent design changes in the library to SMC but that's totally fine, discard my earlier comment we can leave this here for now.

**Comment by rlouf** [comment]
In `rmh.py`: I think `rmh_with_proposal_generator` should be called `rmh`, and move `proposal_generator` that's inside the current `rmh` function to `kenel` . We'll the revisit the general design in #162, and this PR.

File: blackjax/mcmc/rmh.py [hunk 10]
```diff
@@ -172,3 +175,36 @@ def kernel(rng_key: PRNGKey, state: RMHState) -> Tuple[RMHState, RMHInfo]:
...
+    """
+    ndim = jnp.ndim(sigma)  # type: ignore[arg-type]
+    shape = jnp.shape(jnp.atleast_1d(sigma))[:1]
+
+    if ndim == 1:
+        dot = jnp.multiply
+    elif ndim == 2:
+        dot = jnp.dot
+    else:
+        raise ValueError
```
Comment (rlouf) [inline blackjax/mcmc/rmh.py:201]: We need tests that cover these two branches.

**Review by rlouf — CHANGES_REQUESTED** [review]

**Comment by rlouf** [comment]
Awesome, thank you!

**Pull request merged**